### PR TITLE
[framework] Support partial voting in aptos_governance module

### PR DESCRIPTION
**Pull request by xindingw**
### Test Plan

Unit testing.
Move e2e testing.

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 1]
```diff
@@ -84,6 +93,16 @@ module aptos_framework::aptos_governance {
         votes: Table<RecordKey, bool>
     }
 
+    /// Records partial voting proposals.
+    struct PartialVotingProposals has key {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:97]: This looks only useful during the transitioning period but useless after once all proposals are "partial voting proposals". Is there a way to avoid needing this new struct?

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 2]
```diff
@@ -211,6 +245,27 @@ module aptos_framework::aptos_governance {
...
+        if (!exists<PartialVotingProposals>(@aptos_framework)) {
+            return false
+        };
+        let partial_voting_proposals = borrow_global<PartialVotingProposals>(@aptos_framework);
+        table::contains(&partial_voting_proposals.proposals, proposal_id)
+    }
+
+    #[view]
+    /// Return voting power usage of a stake pool on a proposal. Only partial voting proposals could be queried.
+    public fun get_voting_power_usage(stake_pool: address, proposal_id: u64): u64 acquires PartialVotingProposals, VotingRecordsV2 {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:259]: nit: voting power usage is not a good name for this. Also, should we just have a single view function that returns a tuple of vote (yes or no) and voting power associated with the vote?

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 3]
```diff
@@ -294,28 +358,76 @@ module aptos_framework::aptos_governance {
...
+        if (!is_partial_voting_proposal) {
+            assert!(
+                !table::contains(&voting_records.votes, record_key),
+                error::invalid_argument(EALREADY_VOTED));
+            table::add(&mut voting_records.votes, record_key, true);
+        };
 
-        let voting_power = get_voting_power(stake_pool);
+        let staking_pool_voting_power = get_voting_power(stake_pool);
+        if (is_partial_voting_proposal) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:418]: This logic can be greatly simplified by always checking voting records (v1) first, then if not present, check voting records v2. Also we can update vote() to dual write to both voting records v1 and v2 to simplify removal in the future (then we can just simply delete voting records v1 write once there are no longer proposals that are on the previous records)

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 4]
```diff
@@ -53,10 +54,18 @@ module aptos_framework::aptos_governance {
...
+    const EVOTING_POWER_OVERFLOW: u64 = 12;
+    /// Partial voting feature hasn't been properly initialized.
+    const EPARTIAL_VOTING_NOT_INITIALIZED: u64 = 13;
+    /// The proposal in the argument is not a partial voting proposal.
+    const ENOT_PARTIAL_VOTING_PROPOSAL: u64 = 14;
 
     /// This matches the same enum const in voting. We have to duplicate it as Move doesn't have support for enums yet.
     const PROPOSAL_STATE_SUCCEEDED: u64 = 1;
 
+    const MAX_U64: u64 = 18446744073709551615;
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:67]: I wonder if we can just move this into the math_u64 module and add a max_value function or something. @gerben-stavenga what do you think?

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 5]
```diff
@@ -211,6 +237,28 @@ module aptos_framework::aptos_governance {
         borrow_global<GovernanceConfig>(@aptos_framework).required_proposer_stake
     }
 
+    #[view]
+    /// Return remaining voting power of a stake pool on a proposal.
+    public fun get_remaining_voting_power(stake_pool: address, proposal_id: u64): u64 acquires VotingRecords, VotingRecordsV2 {
+        if (features::partial_governance_voting_enabled()) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:243]: I think it's cleaner to create a function is_partial_voting_enabled() which checks both the feature flag and that VotingRecordsV2 is initialized

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 6]
```diff
@@ -211,6 +237,28 @@ module aptos_framework::aptos_governance {
...
+            assert!(exists<VotingRecordsV2>(@aptos_framework), error::invalid_state(EPARTIAL_VOTING_NOT_INITIALIZED));
+        };
+        let record_key = RecordKey {
+            proposal_id,
+            stake_pool,
+        };
+        let voting_records = borrow_global<VotingRecords>(@aptos_framework);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:250]: Let's just add a getter function for this as well has_entirely_voted or something like that that returns true if a stake pool has voted as tracked by VotingRecords (v1)

Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:250]: Can you add some comments for the flow here just for posterity?

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 7]
```diff
@@ -294,28 +342,43 @@ module aptos_framework::aptos_governance {
...
+    ) acquires ApprovedExecutionHashes, GovernanceEvents, VotingRecords, VotingRecordsV2 {
         let voter_address = signer::address_of(voter);
         assert!(stake::get_delegated_voter(stake_pool) == voter_address, error::invalid_argument(ENOT_DELEGATED_VOTER));
 
-        // Ensure the voter doesn't double vote with the same stake pool.
-        let voting_records = borrow_global_mut<VotingRecords>(@aptos_framework);
-        let record_key = RecordKey {
-            stake_pool,
-            proposal_id,
+        let staking_pool_voting_power = get_remaining_voting_power(stake_pool, proposal_id);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:378]: This is a bit strange to me 🤔 Why not do the math forward - if current voted power + new vote > total stake pool's voting power?

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 8]
```diff
@@ -294,28 +342,43 @@ module aptos_framework::aptos_governance {
...
+        vote_internal(voter, stake_pool, proposal_id, voting_power, should_pass);
+    }
+
+    /// Vote on proposal with `proposal_id` and specified voting_power from `stake_pool`.
+    /// If voting_power is more than all the left voting power of `stake_pool`, use all the left voting power.
+    fun vote_internal(
+        voter: &signer,
+        stake_pool: address,
+        proposal_id: u64,
+        voting_power: u64,
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:372]: What happens when voting_power is MAX_U64 here?

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 9]
```diff
@@ -334,6 +397,22 @@ module aptos_framework::aptos_governance {
...
 
+        let record_key = RecordKey {
+            stake_pool,
+            proposal_id,
+        };
+        if (features::partial_governance_voting_enabled()) {
+            let voting_records_v2 = borrow_global_mut<VotingRecordsV2>(@aptos_framework);
+            let used_voting_power = table::borrow_mut_with_default(&mut voting_records_v2.votes, record_key, 0);
+            *used_voting_power = *used_voting_power + voting_power;
+        } else {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:408]: This logic is a bit hard to follow in the case of a proposal created before partial voting is enabled and with voting extending beyond. IMO, it's likely cleaner to fail early if VotingRecords is already updated and:
1. Update VotingRecords if voting power is MAX_U64
2. Update VotingRecordsV2 if partial voting is enabled, regardless of if voting power is MAX_U64 or not.

**Review by movekevin — COMMENTED** [review]
I think there might be some quirks in the user flow here we need to think through: partial_vote is incremental. How would a voter updates their total vote? For example:
1. I voted with 100 APT for proposal 10, which is the total stake I have
2. I added more stake and there's still time left to vote. My stake pool now have 20 APT
3. I want to update the vote to be 20 APT, but now I have to do the math and know to do partial_vote(10)

Furthermore, after partial voting is enabled, if someone still uses the old vote() function, their partial vote records will be updated to MAX_U64, which if I understand the code correctly, would prevent any further voting.

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move (hunk 4 above)
Comment (xindingw) [reply aptos-move/framework/aptos-framework/sources/aptos_governance.move:67]: Sounds good. Since this refactor could affect many places, I will do it in a later PR. Created an issue: https://github.com/aptos-labs/aptos-core/issues/7894

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move (hunk 8 above)
Comment (xindingw) [reply aptos-move/framework/aptos-framework/sources/aptos_governance.move:372]: `voting_power` is bounded by `get_remaining_voting_power` in line 378.

**Comment by xindingw** [comment]
> I think there might be some quirks in the user flow here we need to think through: partial_vote is incremental. How would a voter updates their total vote? For example:

> 1. I voted with 100 APT for proposal 10, which is the total stake I have
> 2. I added more stake and there's still time left to vote. My stake pool now have 20 APT
> 3. I want to update the vote to be 20 APT, but now I have to do the math and know to do partial_vote(10)

What does 10 in `partial_vote(10)` mean here? The user can call `get_remaining_voting_power` to know how much voting power is left and the function will return 20 at 3.

> Furthermore, after partial voting is enabled, if someone still uses the old vote() function, their partial vote records will be updated to MAX_U64, which if I understand the code correctly, would prevent any further voting.

If someone calls the old vote(), only all the current voting power is used.

**Comment by movekevin** [comment]
> > I think there might be some quirks in the user flow here we need to think through: partial_vote is incremental. How would a voter updates their total vote? For example:
> 
> > 1. I voted with 100 APT for proposal 10, which is the total stake I have
> > 2. I added more stake and there's still time left to vote. My stake pool now have 20 APT
> > 3. I want to update the vote to be 20 APT, but now I have to do the math and know to do partial_vote(10)
> 
> What does 10 in `partial_vote(10)` mean here? The user can call `get_remaining_voting_power` to know how much voting power is left and the function will return 20 at 3.
> 
> > Furthermore, after partial voting is enabled, if someone still uses the old vote() function, their partial vote records will be updated to MAX_U64, which if I understand the code correctly, would prevent any further voting.
> 
> If someone calls the old vote(), only all the current voting power is used.

1. The user needing to call get_remaining_vote and also another view function likely to see how much they have voted seems inconvenient. We could introduce a function that allows voting with absolute voting power but I guess that'd introduce complexity if we want to maintain the invariant that voters cannot change their existing vote (e.g. from yes to no).
2. Using all current voting power is fine but not being to add more vote further might not be (since VotingRecords v1 is modified). Not being able to add votes is a very inconsistent/confusing behavior after we enable partial voting.

Ideally, after the update, we should have the following 3 separate voting features:
1. partial_vote for incremental votes (10 more yes votes). This will mainly be used by delegation, liquid staking services, and other pool-based staking protocols.
2. vote() which votes with all current voting power. This should not block adding more votes later
3. vote_total(total_voting_power) to allow voting with an absolute number. This needs more thinking though

**Review by movekevin — COMMENTED** [review]
Are you planning to update the delegation contract in a separate PR to support voting?

**Comment by xindingw** [comment]
> Are you planning to update the delegation contract in a separate PR to support voting?

Yes, splitting into 2 PRs is cleaner and easier for review. That PR is half way done. Letting a delegator delegate its voting power is a bit tricky. Need some product inputs then it should be quick.

**Comment by xindingw** [comment]
> > > I think there might be some quirks in the user flow here we need to think through: partial_vote is incremental. How would a voter updates their total vote? For example:
> > 
> > 
> > > 1. I voted with 100 APT for proposal 10, which is the total stake I have
> > > 2. I added more stake and there's still time left to vote. My stake pool now have 20 APT
> > > 3. I want to update the vote to be 20 APT, but now I have to do the math and know to do partial_vote(10)
> > 
> > 
> > What does 10 in `partial_vote(10)` mean here? The user can call `get_remaining_voting_power` to know how much voting power is left and the function will return 20 at 3.
> > > Furthermore, after partial voting is enabled, if someone still uses the old vote() function, their partial vote records will be updated to MAX_U64, which if I understand the code correctly, would prevent any further voting.
> > 
> > 
> > If someone calls the old vote(), only all the current voting power is used.
> 
> 1. The user needing to call get_remaining_vote and also another view function likely to see how much they have voted seems inconvenient. We could introduce a function that allows voting with absolute voting power but I guess that'd introduce complexity if we want to maintain the invariant that voters cannot change their existing vote (e.g. from yes to no).
> 2. Using all current voting power is fine but not being to add more vote further might not be (since VotingRecords v1 is modified). Not being able to add votes is a very inconsistent/confusing behavior after we enable partial voting.
> 
> Ideally, after the update, we should have the following 3 separate voting features:
> 
> 1. partial_vote for incremental votes (10 more yes votes). This will mainly be used by delegation, liquid staking services, and other pool-based staking protocols.
> 2. vote() which votes with all current voting power. This should not block adding more votes later

The current `vote` doesn't block adding more votes later. `vote` will only mark all the current voting power(instead of MAX_UINT64) is used in VotingRecordsV2. The only case blocking adding more votes is that a stake pool has already voted before we enable partial voting.

> 3. vote_total(total_voting_power) to allow voting with an absolute number. This needs more thinking though
I think we just need to provide a view function for user to get the amount of used voting power? I didn't add it because `VotingRecords` doesn't track it. We can add it later if necessary.
